### PR TITLE
Edit recipe

### PR DESCRIPTION
**Pull request by prashanthich**
Screen Shot 2021-03-31 at 9 33 28 AM
Screen Shot 2021-03-31 at 9 40 11 AM
Screen Shot 2021-03-31 at 9 40 41 AM
Screen Shot 2021-03-31 at 9 40 57 AM

Edit recipe
Delete recipe
view Recipe
Add new category

### CODE REVIEW

File: src/main/java/org/launhcode/healthynutrition/controllers/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    public String displayCategoryForm(Model model){
+        model.addAttribute("title", "Add Category");
+        model.addAttribute(new RecipeCategory());
+        return "recipeCategory/add";
+    }
+
+
+
+    @PostMapping("add")
+    public String processCategoryForm(@ModelAttribute @Valid RecipeCategory recipeCategory,Errors errors,Model model){
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/CategoryController.java:31]: minor: add a space after the commas

File: src/main/java/org/launhcode/healthynutrition/controllers/CategoryController.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    @Autowired
+    private RecipeCategoryRepository recipeCategoryRepository;
+
+    @GetMapping("add")
+    public String displayCategoryForm(Model model){
+        model.addAttribute("title", "Add Category");
+        model.addAttribute(new RecipeCategory());
+        return "recipeCategory/add";
+    }
+
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/CategoryController.java:27]: minor: there should only be one space between these methods

File: src/main/java/org/launhcode/healthynutrition/controllers/RecipeController.java [hunk 3]
```diff
@@ -26,25 +27,70 @@
...
         model.addAttribute("title", "Create Recipe");
+        model.addAttribute("categoryError", "");
         model.addAttribute("categories", recipeCategoryRepository.findAll());
         model.addAttribute(new Recipe());
         return "recipe/create";
     }
 
     @PostMapping("create")
-    public String processCreateRecipeForm(@ModelAttribute @Valid Recipe newRecipe,
+    public String processCreateOrEditRecipeForm(@ModelAttribute @Valid Recipe recipe,
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/RecipeController.java:37]: great variable names! 💯

File: src/main/java/org/launhcode/healthynutrition/controllers/RecipeController.java [hunk 4]
```diff
@@ -26,25 +27,70 @@
...
                                     Errors errors, Model model,
                                     @RequestParam(required = false) List<Integer> categories) {
-        if (errors.hasErrors() || categories.isEmpty() ) {
+
+        if (errors.hasErrors() || categories == null) {
             model.addAttribute("title", "Create Recipe");
+            if (recipe.getId() > 0){
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/controllers/RecipeController.java:43]: Just a question for the team: does everyone understand this check here?

File: src/main/java/org/launhcode/healthynutrition/models/Recipe.java [hunk 5]
```diff
@@ -52,6 +52,7 @@ public Recipe(String name, String description, int calories, String directions,
     public int getId() {
         return id;
     }
+    public void setId(int id){this.id = id;}
```
Comment (korina0430) [inline src/main/java/org/launhcode/healthynutrition/models/Recipe.java:55]: hmm I don't think you should have this one...

**Review by korina0430 — COMMENTED** [review]
Looks awesome! Only thing I think you need to remove is the setId()

**Review by Dhwani0712 — APPROVED** [review]
Looks good!

File: src/main/java/org/launhcode/healthynutrition/controllers/CategoryController.java (hunk 1 above)
Comment (prashanthich) [reply src/main/java/org/launhcode/healthynutrition/controllers/CategoryController.java:31]: Sure, I will do it

File: src/main/java/org/launhcode/healthynutrition/controllers/CategoryController.java (hunk 2 above)
Comment (prashanthich) [reply src/main/java/org/launhcode/healthynutrition/controllers/CategoryController.java:27]: Sure, I will take care

File: src/main/java/org/launhcode/healthynutrition/models/Recipe.java (hunk 5 above)
Comment (prashanthich) [reply src/main/java/org/launhcode/healthynutrition/models/Recipe.java:55]: Without setting ID, I am not able to get recipe id, when I am editing a recipe

Comment (korina0430) [reply src/main/java/org/launhcode/healthynutrition/models/Recipe.java:55]: Hmm that doesn't make sense to me. You said you need to _get_ the recipe id. I don't see where you are setting it?

Comment (korina0430) [reply src/main/java/org/launhcode/healthynutrition/models/Recipe.java:55]: only the db should set this field

**Pull request merged**